doc/export.conf.5
.\" Process this file with
.\" groff -man -Tascii export.conf.5
.\"
.TH EXPORT.CONF 5 "APRIL 2014" RozoFS "User Manuals"
.SH NAME
export.conf \- RozoFS export daemon configuration file
.SH DESCRIPTION
.B export.conf
is manage throw 
.B libconfig
<http://www.hyperrealm.com/libconfig/>
and thus complies to the
.B libconfig
grammar.

.SS layout
    It's the default transform layout used by RozoFS for each volume.
    Each file data will be transform in *forward* chunks, only *inverse* of them
    will be need to reconstruct data (read data), and they will be distributed
    over *safe* storages.

       Valid layouts are:
           - 0 (inverse = 2, forward = 3, safe = 4);
           - 1 (inverse = 4, forward = 6, safe = 8);
           - 2 (inverse = 8, forward = 12, safe = 16);


.SS volumes
   It's a list of volume. Each volume is identified by its vid and may contain
   a list of one or more clusters (cids). Each clusters is identified by its
   cid and may contain at least a list of *safe* storages of the same 
   capacity (sids).

       vid: (volume identifier)
               - must be an integer.
               - must be unique for each volume managed by this exportd.

       layout: (specific layout for this volume / optional)
               - Valid layouts are 0, 1 or 2.
               - Optional field (if no layout is specified for this volume, the
                 default layout is used instead).

       cid: (cluster identifier)
               - must be an integer.
               - must be unique for each cluster managed by this exportd.

       sid: (storage identifier)
               - must be an integer (between 1 and 255).
               - must be unique within a cluster.
	       
       georep: (whether geo-replication must occur on this volume / optional)
               - must be a boolean.
               - Optional field (default value is False).
	       
       host: (IP or hostname of the physical node that contains the storage)
               - must be a valid IPv4 address or a valid hostname.
               - must only be set in case of a volume without geo-replication. 
	       
       site0: (IP or hostname of the physical node that contains the storage on the main site)
               - must be a valid IPv4 address or a valid hostname.
               - must only be set in case of a volume with geo-replication. 	       
	       
       site1: (IP or hostname of the physical node that contains the storage on the secondary site)
               - must be a valid IPv4 address or a valid hostname.
               - must only be set in case of a volume with geo-replication. 	       

.B  WARNING:
   - Use a new cluster when scaling with storage of different capacity.
   - Modifying or removing sid of storage can (will) lead to DATA LOSS !!!
   - Modifying or removing cid can (will) lead to DATA LOSS !!!

.SS exports
   It's the list of rozo file system exported by the exportd daemon. Each
   exported file system is identified by its eid. root is the directory path
   where the metadata for this file system are stored. md5 is a 22 characters
   md5 crypt of the password required to mount the exported filesystem.
   squota is the number of data blocks that can be stored on a file system
   before being notified. hquota is the number maximum of data blocks that can
   be stored on a file system. vid is the volume where the file data for this
   exported filesystem will be stored. bsize defines the block size to be used 
   on this export.

       eid: (export identifier)
               - must be an integer.
               - must be unique for each exported filesystem managed by this
                 exportd.

       bsize: (defines the block size used on this export)
               - Valid bsize are within "4K", "8K", "16K", "32K".
               - default bsize is "4K".

       root: (directory path where metadata for this filesystem will be stored)
               - must be an existing directory.
               - must be unique for each filesystem managed by this exportd.

       md5: (22 characters md5 crypt of the password)
               - it could be generated with: md5 utility (see md5pass (1))
                 the salt used should be: rozofs. Only the last 22 characters
                 should be used (omitting newline).

                 Example: 
                           for password "mypass":
                           md5 is: md5pass mypass rozofs | cut \-c 11\-
                           that is: AyBvjVmNoKAkLQwNa2c4b0

               - For no authentication use empty md5.

       squota: (soft quota)
               - squota stands for soft quota which are non blocking quotas
                 i.e write operations will succeeded but exceeded quotas will
                 be notified in /var/run/exportd/export_[eid] monitoring file.
               - for no soft quota use empty squota.

       hquota: (hard quota)
               - hquota stands for blocking quotas which are blocking quotas
                 i.e every attempt to write beyond hquota will fail (EDQUOT).
               - for no hard quota use empty hquota.

       for squota and hquota:
               - quotas are express in nb. of blocks (no suffix), Kilo, Mega,
                 Giga or Tera Bytes according to suffix K, M, G or T. Only the
                  first letter of  suffix is used that is: 128G is equivalent to
                 128Giga which in turn can be 128GigaBytes etc...
               - warning: any other suffix leads to quota express in blocks.

       vid: (volume identifier used to store file data for this filesystem)
               - must be an integer.
               - should exist in volume list.
               - several filesystem can use the same volume.

.SH EXAMPLE
.PP
.nf
.ta +3i

# sample /etc/rozofs/export.conf file

layout = 0;

volumes =  # List of volumes
(
    {
        # First volume
        vid = 1; # Volume identifier = 1
        cids=    # List of clusters for the volume 1
        (
            {
                # First cluster of volume 1
                cid = 1; # Cluster identifier = 1
                sids =   # List of storages for the cluster 1
                (
                    {sid = 01; host = "storage-node-1-1";},
                    {sid = 02; host = "storage-node-1-2";},
                    {sid = 03; host = "storage-node-1-3";},
                    {sid = 04; host = "storage-node-1-4";}
                    # ...
                );
            },
            {
                 # Second cluster of volume 1
                cid = 2; # Cluster identifier = 2
                sids =   # List of storages for the cluster 2
                (
                    {sid = 01; host = "storage-node-2-1";},
                    {sid = 02; host = "storage-node-2-2";},
                    {sid = 03; host = "storage-node-2-3";},
                    {sid = 04; host = "storage-node-2-4";}
                    # ...
                );
            }
        );
    },
    {
        # Second volume
        vid = 2; # Volume identifier = 2
        georep = True; # Volume is georeplicated 
        layout = 1; # Specific layout for this volume
        cids =   # List of clusters for the volume 2
        (
            {
                # First cluster of volume 2
                cid = 3; # Cluster identifier = 3
                sids =   # List of storages for the cluster 3
                (
                    {sid = 01; site0 = "storage-node-3-1-0"; site1 = "storage-node-3-1-1";},
                    {sid = 02; site0 = "storage-node-3-2-0"; site1 = "storage-node-3-2-1";},
                    {sid = 03; site0 = "storage-node-3-3-0"; site1 = "storage-node-3-3-1";},
                    {sid = 04; site0 = "storage-node-3-4-0"; site1 = "storage-node-3-4-1";},
                    {sid = 05; site0 = "storage-node-3-5-0"; site1 = "storage-node-3-5-1";},
                    {sid = 06; site0 = "storage-node-3-6-0"; site1 = "storage-node-3-6-1";},
                    {sid = 07; site0 = "storage-node-3-7-0"; site1 = "storage-node-3-7-1";},
                    {sid = 08; site0 = "storage-node-3-8-0"; site1 = "storage-node-3-8-1";}
                );
            }
        );
    }
)
;

exports = ( # List of exported filesystem

    # First filesystem exported
    {eid = 1; bsize="4K"; root = "/path/to/foo"; md5="AyBvjVmNoKAkLQwNa2c4b0";
                                    squota="128G"; hquota="256G"; vid=1;},

    # Second filesystem exported
    {eid = 2; bsize="8K"; root = "/path/to/bar"; md5="";
                            squota=""; hquota = ""; vid=2;}
    # ...
);

.SH FILES
.I /etc/rozofs/export.conf (/usr/local/etc/rozofs/export.conf)
.RS
The system wide configuration file.
.\".SH ENVIRONMENT
.\".SH DIAGNOSTICS
.\".SH BUGS
.SH AUTHOR
Fizians <http://www.fizians.com>
.SH "SEE ALSO"
.BR rozofs (7),
.BR exportd (8)


debian/changelog
rozofs (2.0.alpha21) stable; urgency=low

  * (rozofs) See git log for changes

 -- Fizians S.A.S. <devel@rozofs.com>  Mon, 01 Jan 2014 00:00:00 +0100


manager/rozofs/cli/export.py
# -*- coding: utf-8 -*-
#
# Copyright (c) 2013 Fizians SAS. <http://www.fizians.com>
# This file is part of Rozofs.
#
# Rozofs is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published
# by the Free Software Foundation, version 2.
#
# Rozofs is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see
# <http://www.gnu.org/licenses/>.

import sys
from rozofs.core.platform import Platform, Role
from rozofs.core.agent import ServiceStatus
from rozofs.cli.output import ordered_puts
from rozofs.cli.output import puts
from collections import OrderedDict


def list(platform, args):
    configurations = platform.get_configurations([args.exportd], Role.EXPORTD)
    if configurations[args.exportd] is None:
        raise Exception("%s is not reachable" % args.exportd)

    configuration = configurations[args.exportd][Role.EXPORTD]

    list_l = {}
    exports_l = []
    for eid, econfig in configuration.exports.items():
        export_l = []
        export_l.append({'vid':econfig.vid})
        export_l.append({'root':econfig.root})
        export_l.append({'md5':econfig.md5})
        export_l.append({'squota':econfig.squota})
        export_l.append({'hquota':econfig.hquota})

        exports_l.append({'EXPORT '+ str(eid):export_l})
    list_l.update({'EXPORTS': exports_l})
    ordered_puts(list_l)

def create(platform, args):
    platform.create_export(args.vid[0], args.name, args.passwd, args.squota, args.hquota)


def update(platform, args):
    platform.update_export(args.eid[0], args.current, args.passwd, args.squota, args.hquota)


def remove(platform, args):
    if not args.eids:
        args.eids = None

    platform.remove_export(args.eids, args.force)

def get(platform, args):
    configurations = platform.get_configurations([args.exportd], Role.EXPORTD)
    if configurations[args.exportd] is None:
        raise Exception("exportd node is off line.")

    configuration = configurations[args.exportd][Role.EXPORTD]

    list_l = {}
    exports_l = []
    for eid in args.eids:
        econfig = configuration.exports[eid]
        export_l = []
        export_l.append({'vid':econfig.vid})
        export_l.append({'root':econfig.root})
        export_l.append({'md5':econfig.md5})
        export_l.append({'squota':econfig.squota})
        export_l.append({'hquota':econfig.hquota})

        exports_l.append({'EXPORT '+ str(eid):export_l})
    list_l.update({'EXPORTS': exports_l})
    ordered_puts(list_l)

def mount(platform, args):
    if not args.eids:
        args.eids = None

    platform.mount_export(args.eids, args.nodes, args.options)


def umount(platform, args):
    if not args.eids:
        args.eids = None

    platform.umount_export(args.eids, args.nodes)


def dispatch(args):
    only_export_actions = {'list', 'create', 'update', 'get'}

    if (args.action in only_export_actions):
        p = Platform(args.exportd, Role.EXPORTD)
    else:
        p = Platform(args.exportd)

    globals()[args.action.replace('-', '_')](p, args)



